features/frameworks/greentea-client/source/test_env.cpp

```diff
@@ -61,8 +61,10 @@ static void greentea_notify_version();
  * Rawserial object used to provide direct, raw serial communications
  * between the target and the host.
  */
-RawSerial greentea_serial(USBTX, USBRX); 
-
+RawSerial& greentea_serial() { 
+    static RawSerial serial(USBTX, USBRX);
```

Comment:
> Hmm, I'm probably missing it, but I'm not sure I see where SingletonPtr benefits over static initialization other than the syntax difference, but both cases are different from the syntax for global objects.

Ok, let me go back to the start. 

When a compilation unit define global (`extern` or `static`) variable(s) that are not initialized at compile time, then the compiler generate an initialization function which will initialize these global variables of this compilation unit before the entry in the `main` function. 

Some examples of *runtime* global: 

```c++

extern "C" { 
int foo();
const int bar = foo();
}

// the definition is omitted but Baz is not a POD 
class Baz;

Baz baz;
``` 

This initialization function will be referenced in special section which contains all the initialization functions of the program. Before the entry in main, the c runtime will execute every functions referenced in this section. 

The main issue with that model is that global variables initialized at runtime can't be removed from the output because even if it seems like the variable is not used, it is used by the c runtime. 

Then the dependency chain is: 

`c runtime` -> `initialization_function_array` -> `compilation unit initialization`  -> `global variable`. 

That was the issue with the `utest` code and what this PR fix. A lot of global variables initialized at runtime were here and these variables, not collectable by the linker GC,  pulled code and data from the utest library.

Global runtime (static or extern) variables are **never** an option for embedded libraries. 

Wrapping the variable inside a `LazyInstance` or a `function scope` (as static) is the only option here (both have their advantages and inconvenient).